daemon: sync BPF maps with in-memory K8s service maps

ianvernon opened the pull request:
This provides a fix for the following scenario when Cilium is running in tandem with Kubernetes:
    
* Cilium is running, and a service is added to Kubernetes. Cilium's Kubernetes watcher receives this event, and plumbs the service into its BPF service maps.
* Cilium is stopped / stops running.
* Aforementioned service is deleted from Kubernetes.
* Cilium is restarted, but the service is kept in the BPF maps because Cilium never received the deletion event for the service from Kubernetes.

Now, upon bootstrapping of the daemon, Cilium waits for all Kubernetes resources that exist upon startup of the Cilium agent that get plumbed into the BPF service maps (services, ingresses) to be consumed by Cilium. Once this is done, a controller is spawned which synchronizes the contents of the BPF service maps with the in-memory Kubernetes service maps maintained by Cilium. This ensures that any service which was deleted when Cilium was not running is cleaned up. Once this controller succeeds, it does not run again; this means that any service which is added manually, say by `cilium service add`, will not be deleted by Cilium if the controller has already completed successfully, or until Cilium is restarted again.

Along the way, did some tidying up and refactoring of some code to allow for reuse of it within this new controller.

Fixes: #4457

Signed-off by: Ian Vernon <ian@cilium.io>

Some comments: I most likely will refactor this in the future so we are doing this on a per-map basis to be in-line with #3161 . Eventually, I would like to factor out the code in such a way that we can spawn controllers on a per-map basis, but this wouldn't do much because we have to hold global locks in the loadbalancer in its current implementation to do this anyway, so I don't think there's not much of a performance benefit by doing this currently. Also would like to get comments regarding running the controller with a `RunInterval` of 0 (meaning it doesn't run again after succeeding) - should we run this consistently to ensure synchronicity between `cilium-agent` and the datapath? Such a change is trivial. I chose not to do it because when this controller is ran, it holds two global mutexes in the loadbalancer. 

```release-note
clean up BPF service map entries for services that were deleted from Kubernetes while Cilium wasn't running
```

**Testing-done**:

I ran this against two test cases in the developer VM:

`NFS=1 K8S=1 ./contrib/vagrant/start.sh`

The following examples were used for testing:

* svc.yaml:

```
apiVersion: v1
kind: Service
metadata:
  name: testds-service
spec:
  ports:
  - port: 80
  selector:
    zgroup: testDS
```

* ds.yaml:

```
apiVersion: extensions/v1beta1
kind: DaemonSet
metadata:
  name: testds
  namespace: default
spec:
  template:
    metadata:
      labels:
        zgroup: testDS
    spec:
      containers:
      - name: web
        image: docker.io/cilium/demo-httpd
        ports:
        - containerPort: 80
      tolerations:
      - effect: NoSchedule
        key: node-role.kubernetes.io/master
      - effect: NoSchedule
        key: node.cloudprovider.kubernetes.io/uninitialized
        value: "true"
```

*No corresponding backends for service exist*

1. List services in Cilium:

```
$ cilium service list
ID   Frontend             Backend                        
1    172.20.0.10:53/UDP   1 => 10.0.96.249:53/NONE       
2    172.20.0.1:443/TCP   1 => 192.168.33.11:6443/NONE 
```

2. BPF map contents match:
```
$ sudo cilium bpf lb list
SERVICE ADDRESS   BACKEND ADDRESS          
172.20.0.1:443    0.0.0.0:0 (0)            
                  192.168.33.11:6443 (2)   
172.20.0.10:53    10.0.96.249:53 (1)       
                  0.0.0.0:0 (0)   
```

3. While Cilium is still running, import the service:

```
$ kubectl apply -f svc.yaml 
service/testds-service created
```

4. Stop cilium:

```
$ sudo service cilium stop
$
```

5. While Cilium is stopped, delete the service:
```
$ kubectl delete -f svc.yaml 
service "testds-service" deleted
```

6. Confirm that service is deleted in K8s:
```
$ kubectl get svc --all-namespaces
NAMESPACE     NAME         TYPE        CLUSTER-IP    EXTERNAL-IP   PORT(S)         AGE
default       kubernetes   ClusterIP   172.20.0.1    <none>        443/TCP         4h
kube-system   kube-dns     ClusterIP   172.20.0.10   <none>        53/UDP,53/TCP   4h
```

7. Confirm that `cilium bpf lb list` still shows the service as being in the BPF maps (this is expected because cilium isn't running):

```
$ sudo cilium bpf lb list
SERVICE ADDRESS   BACKEND ADDRESS          
172.20.0.10:53    10.0.96.249:53 (1)       
                  0.0.0.0:0 (0)            
172.20.0.43:80    0.0.0.0:0 (0)            
172.20.0.1:443    0.0.0.0:0 (0)            
                  192.168.33.11:6443 (2)
```

8. Start Cilium:
```
$ sudo service cilium start
```

9. Check that the service is removed from the the service list / BPF maps:
```
$ sudo cilium bpf lb list
SERVICE ADDRESS   BACKEND ADDRESS          
172.20.0.1:443    0.0.0.0:0 (0)            
                  192.168.33.11:6443 (1)   
172.20.0.10:53    10.0.96.249:53 (2)       
                  0.0.0.0:0 (0)       
```

```
$ cilium service list
ID   Frontend             Backend                        
1    172.20.0.1:443/TCP   1 => 192.168.33.11:6443/NONE   
2    172.20.0.10:53/UDP   1 => 10.0.96.249:53/NONE 
```

*Backends exist for service* 

1. Apply DaemonSet and service:

```
$ kubectl apply -f ds.yaml 
daemonset.extensions/testds created
$ kubectl apply -f svc.yaml 
service/testds-service created
```

2. Confirm service list and BPF map list show new service:
```
$ cilium service list
ID   Frontend             Backend                        
1    172.20.0.1:443/TCP   1 => 192.168.33.11:6443/NONE   
2    172.20.0.10:53/UDP   1 => 10.0.96.249:53/NONE       
3    172.20.0.75:80/TCP   1 => 10.0.174.167:80/NONE 
```

```
$ sudo cilium bpf lb list 
SERVICE ADDRESS   BACKEND ADDRESS          
172.20.0.1:443    0.0.0.0:0 (0)            
                  192.168.33.11:6443 (1)   
172.20.0.75:80    0.0.0.0:0 (0)            
                  10.0.174.167:80 (3)      
172.20.0.10:53    10.0.96.249:53 (2)       
                  0.0.0.0:0 (0)          
```

4. Stop Cilium

```
$ sudo service cilium stop
$
```

5. Delete service:
```
$ kubectl delete -f svc.yaml 
service "testds-service" deleted
```

6. Start CIlium:
```
$ sudo service cilium start
$
```

7.  Service is not in BPF lb list or in service list:

```
$ sudo cilium bpf lb list 
SERVICE ADDRESS   BACKEND ADDRESS          
172.20.0.1:443    0.0.0.0:0 (0)            
                  192.168.33.11:6443 (2)   
172.20.0.10:53    10.0.96.249:53 (1)       
                  0.0.0.0:0 (0)      
```

```
$ cilium service list
ID   Frontend             Backend                        
1    172.20.0.10:53/TCP   1 => 10.0.96.249:53/NONE       
2    172.20.0.1:443/TCP   1 => 192.168.33.11:6443/NONE 
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5203)
<!-- Reviewable:end -->

Commit:
daemon: sync BPF maps with in-memory K8s service maps

This provides a fix for the following scenario when Cilium is running in tandem
with Kubernetes:

* Cilium is running, and a service is added to Kubernetes. Cilium's Kubernetes
watcher receives this event, and plumbs the service into its BPF service maps.
* Cilium is stopped / stops running.
* Aforementioned service is deleted from Kubernetes.
* Cilium is restarted, but the service is kept in the BPF maps because Cilium
never received the deletion event for the service from Kubernetes.

Now, upon bootstrapping of the daemon, Cilium waits for all Kubernetes resources
that exist upon startup of the Cilium agent that get plumbed into the BPF
service maps (services, ingresses) to be consumed by Cilium. Once this is done,
a controller is spawned which synchronizes the contents of the BPF service maps
with the in-memory Kubernetes service maps maintained by Cilium. This ensures
that any service which was deleted when Cilium was not running is cleaned up.
Once this controller succeeds, it does not run again; this means that any
service which is added manually, say by `cilium service add`, will not be
deleted by Cilium if the controller has already completed successfully, or until
Cilium is restarted again.

Fixes: #4457

Signed-off by: Ian Vernon <ian@cilium.io>

ianvernon commented on daemon/daemon.go:
```diff
@@ -145,10 +145,10 @@ type Daemon struct {
 	// k8sResourceSyncWaitGroup is used to block the starting of the daemon,
 	// including regenerating restored endpoints (if specified) until all
-	// policies stored in Kubernetes are plumbed into the local Cilium
+	// policies / servicesstored in Kubernetes are plumbed into the local Cilium
```
Will fix this typo.

ianvernon commented on daemon/daemon.go:
```diff
@@ -145,10 +145,10 @@ type Daemon struct {
 	// repository.
-	// This prevents regeneration of endpoints before all policy rules in
-	// Kubernetes are consumed by Cilium. See GH-5038.
+	// This prevents regeneration of endpoints before all policy rules / services
```
Also mention Ingresses, as these update the loadbalancer maps.

ianvernon commented on daemon/main.go:
```diff
@@ -825,16 +825,16 @@ func runDaemon() {
 
 	go func() {
-		log.Info("Waiting until all pre-existing policies have been received")
+		log.Info("Waiting until all pre-existing policies and services have been received")
```
also add ingresses

ianvernon commented on daemon/main.go:
```diff
@@ -825,16 +825,16 @@ func runDaemon() {
 	select {
 	case <-cachesSynced:
-		log.Info("All pre-existing policies have been received; continuing")
+		log.Info("All pre-existing policies and services have been received; continuing")
```
also add ingresses

ianvernon commented on daemon/main.go:
```diff
@@ -843,6 +843,21 @@ func runDaemon() {
 		d.regenerateRestoredEndpoints(restoredEndpoints)
 		go func() {
+			if k8s.IsEnabled() {
+				log.Debugf("waiting for all services to be listed by Kubernetes before syncing loadbalancer maps")
```
Remove these logs; we are guaranteed to be synced by this point.

ianvernon commented on daemon/main.go:
```diff
@@ -843,6 +843,21 @@ func runDaemon() {
+					},
+				)
+				return
+			}
```
Note that we do not call `SyncLBMap` when running with K8s. I believe that this is fine, primarily because we are trusting K8s to provide us with the information we need instead of the existing state in the datapath; the cleanup happens after we get the services from K8s, so there is no worry about having the maps temporarily cleared of any entry at any given time.

Commit:
pkg/loadbalancer: add StringWithProtocol to L3n4Addr

The format is the following for the string representation of L3n4Addr:

* IPv4: IP:Port/Protocol
* IPv6: [IP]:Port/Protocol

This ensures that the string representation of the L3n4Addr is unique from any
other L3n4Addr. This will be used in an upcoming commit which stores L3n4Addrs
in a set.

Fixes: #4457

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer commented on daemon/daemon.go:
```diff
@@ -145,10 +145,11 @@ type Daemon struct {
-	// repository.
-	// This prevents regeneration of endpoints before all policy rules in
-	// Kubernetes are consumed by Cilium. See GH-5038.
+	// policies, services, and ingresses stored in Kubernetes at the time of
```
At a glance I was confused by "ingresses", I assume these are (kubernetes) "ingress resources"? Perhaps it would be a little more explicit to refer to them like this?

ianvernon replied:
I personally thought that "policies, services, and ingresses stored in Kubernetes" would elucidate that an ingress is an object in Kubernetes.

joestringer commented on daemon/k8s_watcher.go:
```diff
@@ -400,6 +400,17 @@ func (d *Daemon) EnableK8sWatcher(reSyncPeriod time.Duration) error {
+			// If sync fails, fatally exit.
+			log.Fatalf("failed to wait for cache to sync for Services")
+		}
+		log.Debug("cache synced for Services")
```
All the other messages say "caches synced ...".

These changes look somewhat orthogonal to the commit's intent, they probably could be split to reduce the changes in each commit (but also if that's too much of a pain it's fine by me to keep them in this commit).

Also, with the two existing instances of creating a k8s resource sync goroutine plus the two being introduced in this patch, it looks like the main thing that changes is the second parameter to `cache.WaitForCacheSync()`. Seems like a prime target for a simple refactor into a function something like:

```
func blockWaitGroupToSyncResource(waitGroup sync.WaitGroup, informer cache.SharedIndexInformer) {
    waitGroup.Add(1)
    go func() {
        ...
    }
}
```

ianvernon replied:
Good idea! I'll add this in my next upload.

joestringer commented on daemon/loadbalancer.go:
```diff
@@ -389,6 +389,17 @@ func (d *Daemon) RevNATDelete(id loadbalancer.ServiceID) error {
 	return err
 }
 
+func (d *Daemon) deleteRevNATBPFLocked(id loadbalancer.ServiceID, isIPv4 bool) error {
```
This doesn't seem to use the daemon at all, and with some minor tweaks looks like it could be pushed into `pkg/loadbalancer/` or `pkg/maps/lbmap`.

ianvernon replied:
Will be in next upload.

joestringer commented on daemon/loadbalancer.go:
```diff
@@ -438,6 +449,100 @@ func (d *Daemon) RevNATDump() ([]loadbalancer.L3n4AddrID, error) {
+// disabled or not for the daemon. If includeMasterBackend is true, the returned
+// values will also include services which correspond to "master" backend values
+// in the BPF maps Returns the errors that occurred while dumping the maps.
+func dumpBPFServiceMapsToUserspace(includeMasterBackend bool) (loadbalancer.SVCMap, []*loadbalancer.LBSVC, []error) {
```
Great to see this factored out from the below function. This also looks like it should be generic enough to live in `pkg/maps/lbmap/`, except maybe for the `option.Config` check.

ianvernon replied:
Good idea. The config check can be factored out into a boolean parameter to said functions. Will be in next upload :D

joestringer commented on daemon/main.go:
```diff
@@ -825,16 +825,16 @@ func runDaemon() {
 
 	go func() {
-		log.Info("Waiting until all pre-existing policies have been received")
+		log.Info("Waiting until all pre-existing Kubernetes resources related to policy have been received")
```
is it related to policy or just generally Kubernetes resources?

ianvernon replied:
We don't wait for all Kubernetes resources, just the ones related to networking (Service, Ingress, NetworkPolicy, Endpoint, and CiliumNetworkPolicy) are synced. I'm open to recommendations about how to make this more clear.

joestringer commented on daemon/loadbalancer.go:
```diff
@@ -438,6 +449,100 @@ func (d *Daemon) RevNATDump() ([]loadbalancer.L3n4AddrID, error) {
+	newRevNATMap := loadbalancer.RevNATMap{}
+	errors := make([]error, 0, 2)
+
+	// TODO cleanup revnat mappings in ref to k8s services??
```
Don't you already do this below?

ianvernon replied:
I'll remove this comment since the concern in the TODO is outside of the purview of this function. But yes, the controller does cleanup revNATs in reference to K8s services now.

joestringer reviewed (approved):
Looks good, I only really made code organisation comments. In particular there seems to be a bunch of refactoring happening in the same patch as functional changes and some of that refactored code could live in packages outside `daemon/`.

Commit:
daemon: factor K8s waitGroup synchronization into function

The same code was being used in multiple places, but for different Kubernetes
controllers, which makes refactoring it to be a function which is provided a
controller an easy, and more maintainable way of writing such code. Replace
existing equivalent code for NetworkPolicy and CiliumNetworkPolicy controllers,
and utilize it for Service, Ingress, and Endpoint controllers as well.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
lbmap: add DeleteRevNATBPF function

Add a function which deletes a specific revNAT entry from its corresponding BPF
map based on if the ID corresponds to the IPv4 or IPv6 revNAT maps. Replace
code which did this already with this function. This function will be used
again in an upcoming commit, which was the primary motivation behind this
refactor.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
refactor dump to userspace representation from daemon to lbmap

Converting from BPF key-value representation to the corresponding userspace
representations for the service and revNAT maps has several uses. Previously,
this was only done in the `SyncLB` function for the daemon. In an upcoming
commit, such functionality will be utilized elsewhere to synchronize the set of
services received by Kubernetes with the BPF maps. Factor out this functionality
into functions within the lbmap function accordingly. As an added bonus, make
some functions in the lbmap package private because they are not needed outside
of the lbmap package anymore after this refactor.

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -548,3 +548,16 @@ func ServiceValue2LBBackEnd(svcKey ServiceKey, svcValue ServiceValue) (*loadbala
+// (IPv4 or IPv6) with ID id. Returns an error if the deletion operation failed.
+func DeleteRevNATBPF(id loadbalancer.ServiceID, isIPv4 bool) error {
+	var revNATK RevNatKey
+	if isIPv4 {
```
Most, if not all functions have an `isIPv6` instead of `isIPv4`. Consider change the function's argument to have `isIPv6` instead.

ianvernon replied:
Fixed in next upload.

ianvernon replied:
Fixed, and ensured that callers were changed to account for this swapping of logic.

aanm commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -561,3 +561,96 @@ func DeleteRevNATBPF(id loadbalancer.ServiceID, isIPv4 bool) error {
+
+		scopedLog.Debug("parsing service mapping")
+		fe, be, err := serviceKeynValue2FEnBE(svcKey, svcValue)
+		if err != nil {
```
Why isn't this error appended to the `errors` slice?

ianvernon replied:
Fixed in next upload.

aanm commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -561,3 +561,96 @@ func DeleteRevNATBPF(id loadbalancer.ServiceID, isIPv4 bool) error {
+
+		scopedLog.Debug("parsing BPF revNAT mapping")
+		fe, err := revNatValue2L3n4AddrID(revNatK, revNatV)
+		if err != nil {
```
Why isn't this error appended to the `errors` slice?

ianvernon replied:
Good idea, I'll add it.

ianvernon replied:
Fixed in next upload.

aanm commented on daemon/loadbalancer.go:
```diff
@@ -597,3 +597,124 @@ func (d *Daemon) SyncLBMap() error {
+		}
+	}
+
+	bpfDeleteErrors := make([]error, 0, len(k8sDeletedServices)+len(k8sDeletedRevNATS))
```
the allocation here is unnecessary as we are not expecting to have errors while doing this operation. In other words you are making the best case scenario slower.
